### PR TITLE
feat: transitive composite action scanning (Phase 2)

### PR DESCRIPTION
**Pull request by kotakanbe**
## Summary

- Add ADR-0008 documenting the decision to implement transitive composite action scanning, superseding ADR-0007 (skip decision)
- Implement recursive BFS resolution of composite action dependencies under `--include-actions --show-transitive`
- Replace section-based display (`--- GitHub Actions ---`) with conditional `SOURCE` column in table/detailed output
- Default `--include-actions` shows direct actions only; `--show-transitive` opts in to transitive output
- Transitive entries include a `Via` field showing which direct action caused the discovery

## Motivation

Composite actions execute their `steps[].uses` on the **caller's runner**, giving transitive dependencies direct access to `GITHUB_TOKEN`, secrets, and source code. However, for **health/lifecycle management** (as opposed to vulnerability management), transitive issues are not directly actionable by the user — if a direct action is healthy, its maintainer is presumably managing their own deps. Therefore transitive results are opt-in.

### `--include-actions` (default: direct only)

```
$ uzomuzo scan https://github.com/future-architect/uzomuzo-oss --include-actions

VERDICT  SOURCE  PURL                                              LIFECYCLE  EOL
ok       direct  https://github.com/future-architect/uzomuzo-oss   Active     Not EOL
ok       action  https://github.com/actions/checkout               Active     Not EOL
ok       action  https://github.com/actions/setup-go               Active     Not EOL
ok       action  https://github.com/actions/upload-artifact        Active     Not EOL
ok       action  https://github.com/anthropics/claude-code-action  Active     Not EOL
ok       action  https://github.com/aquasecurity/trivy-action      Active     Not EOL
ok       action  https://github.com/github/codeql-action           Active     Not EOL
ok       action  https://github.com/golangci/golangci-lint-action  Active     Not EOL
ok       action  https://github.com/goreleaser/goreleaser-action   Active     Not EOL
ok       action  https://github.com/ossf/scorecard-action          Active     Not EOL
ok       action  https://github.com/sigstore/cosign-installer      Active     Not EOL

Summary: 11 dependencies | 11 ok | 0 caution | 0 replace | 0 review
```

### `--include-actions --show-transitive` (with transitive + Via)

```
$ uzomuzo scan https://github.com/future-architect/uzomuzo-oss --include-actions --show-transitive

VERDICT  SOURCE             PURL                                              LIFECYCLE  EOL
ok       direct             https://github.com/future-architect/uzomuzo-oss   Active     Not EOL
ok       action             https://github.com/actions/checkout               Active     Not EOL
...
ok       action             https://github.com/sigstore/cosign-installer      Active     Not EOL
ok       action-transitive  https://github.com/actions/cache                  Active     Not EOL
ok       action-transitive  https://github.com/aquasecurity/setup-trivy       Active     Not EOL
ok       action-transitive  https://github.com/oven-sh/setup-bun              Active     Not EOL

Summary: 14 dependencies | 14 ok | 0 caution | 0 replace | 0 review
```

3 transitive actions discovered via BFS composite resolution.

### `--format detailed --show-transitive` (transitive entries with Via)

```
--- PURL 12 (action-transitive) ---
🔗 Via: https://github.com/aquasecurity/trivy-action
📦 Package: https://github.com/actions/cache
🧾 Description: Cache dependencies and build outputs in GitHub Actions
⚖️  Result: 🟢 Active
💭 Reason: Recent human commits but no recent package publishing
📊 GitHub Info: Normal (⭐ 5324 stars)
📄 Licenses: Project: MIT
💻 Latest Commit: 2026-03-18
🔗 Repository: https://github.com/actions/cache

--- PURL 13 (action-transitive) ---
🔗 Via: https://github.com/aquasecurity/trivy-action
📦 Package: https://github.com/aquasecurity/setup-trivy
🧾 Description: Set up your GitHub Actions workflow with a specific version of Trivy
⚖️  Result: 🟢 Active
📊 GitHub Info: Normal (⭐ 11 stars)
📄 Licenses: Project: Apache-2.0
💻 Latest Commit: 2026-01-15
🔗 Repository: https://github.com/aquasecurity/setup-trivy

--- PURL 14 (action-transitive) ---
🔗 Via: https://github.com/anthropics/claude-code-action
📦 Package: https://github.com/oven-sh/setup-bun
🧾 Description: Set up your GitHub Actions workflow with a specific version of Bun
⚖️  Result: 🟢 Active
📊 GitHub Info: Normal (⭐ 688 stars)
📄 Licenses: Project: MIT
💻 Latest Commit: 2026-03-14
🔗 Repository: https://github.com/oven-sh/setup-bun
```

### `--format json --show-transitive` (transitive entries with via field)

```json
{
  "purl": "https://github.com/actions/cache",
  "verdict": "ok",
  "lifecycle": "Active",
  "eol": "Not EOL",
  "repo_url": "https://github.com/actions/cache",
  "project_license": "MIT",
  "source": "actions-transitive",
  "via": "https://github.com/aquasecurity/trivy-action"
}
{
  "purl": "https://github.com/aquasecurity/setup-trivy",
  "verdict": "ok",
  "source": "actions-transitive",
  "via": "https://github.com/aquasecurity/trivy-action"
}
{
  "purl": "https://github.com/oven-sh/setup-bun",
  "verdict": "ok",
  "source": "actions-transitive",
  "via": "https://github.com/anthropics/claude-code-action"
}
```

## Changes by layer

| Layer | Change |
|-------|--------|
| Domain | `SourceActionsTransitive`, `Via string` on `AuditEntry` |
| Infrastructure | `ActionRef`, `ParseCompositeActionURLs`, BFS with parent tracking (`queueItem.via`), `fetchActionYAML`, `DiscoverActions` returns `map[string]string` for transitive |
| Application | `ActionsDiscoverer` interface updated; `ActionsConfig.ShowTransitive`; `evaluateActionURLs`; Via population from discovery map |
| Interfaces | `--show-transitive` flag; conditional `SOURCE` column; `🔗 Via:` in detailed; `via` in JSON/CSV |

## Design decisions

- **Direct-only by default** — health management focuses on actionable results; transitive health is the upstream maintainer's responsibility
- **`--show-transitive` is generic** — designed to also control future SBOM transitive dependency display
- **Via shows the direct action, not intermediate** — for A→B→C, C's Via is A (what the user chose and can replace)
- **First-discovered parent wins** — when multiple direct actions lead to the same transitive dep, the lexicographically first parent is shown (deterministic)
- **BFS skipped when not needed** — no API calls for transitive resolution unless `--show-transitive` is set
- **Conditional SOURCE column** — only shown when entries have multiple source types

## Test plan

- [x] `TestExtractActionRef` — all edge cases (subdirectory, SHA-pinned, local, docker)
- [x] `TestParseCompositeActionURLs` — composite and non-composite actions
- [x] `TestActionRef_ActionYAMLPath` — root and subdirectory path construction
- [x] `TestRenderScanTable_WithSourceColumn` — SOURCE column with mixed sources
- [x] `TestRenderScanTable_NoSourceColumnForSingleSource` — no column for uniform source
- [x] `TestRenderScanJSON_WithSource` — JSON includes `actions-transitive` source
- [x] `TestRenderScanCSV_WithSource` — CSV includes `via` column
- [x] `go test ./internal/...` — all pass
- [x] `go vet ./...` — clean
- [x] Real execution: direct-only → 11 deps; with `--show-transitive` → 14 deps + Via annotation

Closes Phase 2 of #98.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by kotakanbe** [comment]
### Copilot Review Resolution (`--copilot-only` mode) 

- [x] Read `.github/prompts/review.prompt.md` for procedure
- [x] Phase 2: Discover and classify all Copilot review comments
- [ ] Phase 2: Apply fixes for FIX-classified comments
- [ ] Phase 2: Build/test/lint
- [ ] Phase 2: Commit and push
- [ ] Phase 2: Reply and resolve threads
- [ ] Phase 3: Detect recurring patterns → propose coding rules
- [ ] Phase 3: Apply rules to `.github/instructions/` and run `make sync-instructions`
- [ ] Phase 4: Post summary comment

### Classifications (2nd Copilot review — 5 unresolved threads)

| # | Thread | File | Classification | Reason |
|---|--------|------|---------------|--------|
| 6 | `PRRT_kwDORi6Hcs54quK-` | `service.go:224` | ALREADY_FIXED | Error is now propagated at line 230 via `fmt.Errorf` |
| 7 | `PRRT_kwDORi6Hcs54quLR` | `discovery.go:139` | ALREADY_FIXED | `buildActionRefFromGitHubURL` preserves subdirectory path; `actionRefKey` includes path |
| 8 | `PRRT_kwDORi6Hcs54quLa` | `discovery.go:166` | ALREADY_FIXED | Error stored directly without re-wrapping (line 204) |
| 9 | `PRRT_kwDORi6Hcs54quLp` | `discovery.go:105` | FIX | Add unit tests for `buildActionRefFromGitHubURL` and `actionRefKey` |
| 10 | `PRRT_kwDORi6Hcs54quL2` | `ADR-0008:100` | FIX | ADR claims source in per-entry header but implementation only does this for error entries; update ADR to match |

[View job run](https://github.com/future-architect/uzomuzo-oss/actions/runs/23937352645)

File: internal/infrastructure/depparser/ghaworkflow/parser.go [hunk 1]
```diff
@@ -102,6 +102,111 @@ func extractGitHubURL(uses string) string {
...
+type actionRuns struct {
+	Using string `yaml:"using"`
+	Steps []step `yaml:"steps"`
+}
+
+// ParseCompositeActionURLs parses an action.yml file and extracts GitHub URLs
+// from steps[].uses if the action is a composite action (runs.using: composite).
+//
+// Returns:
+//   - refs: ActionRef values for each uses: directive found in composite steps
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/infrastructure/depparser/ghaworkflow/parser.go]: Fixed in 135d014.

File: internal/infrastructure/actionscan/discovery.go [hunk 2]
```diff
@@ -89,7 +88,7 @@ func (s *DiscoveryService) DiscoverActions(ctx context.Context, repoURLs []strin
 			for _, u := range urls {
 				if _, exists := result.Actions[u]; !exists {
 					result.Actions[u] = 0
-					orderedURLs = append(orderedURLs, u)
+					directURLs = append(directURLs, u)
 				}
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/infrastructure/actionscan/discovery.go]: Not applicable — same issue raised and rejected by maintainer in prior review rounds (threads #4, #7, #11). `buildActionRefFromGitHubURL` preserves subdirectory paths from URLs, and `actionRefKey` includes the path for BFS dedup.

**Comment by kotakanbe** [comment]
@claude Follow the procedure in `.github/prompts/review.prompt.md` with `--copilot-only` mode for this PR.

Phase 2: Resolve all unresolved Copilot review threads (discover → classify → fix → build/test/lint → commit/push → reply/resolve).
Phase 3: Detect recurring patterns from FIX-classified comments, propose and apply coding rules to `.github/instructions/`, then run `make sync-instructions`.
Phase 4: Post a summary comment on this PR.

File: internal/infrastructure/depparser/ghaworkflow/parser.go [hunk 3]
```diff
@@ -102,6 +102,111 @@ func extractGitHubURL(uses string) string {
...
+//
+// Returns:
+//   - refs: parsed ActionRef values (owner/repo/path/ref) for each uses: directive found in composite steps
+//   - isComposite: true if runs.using == "composite"
+//   - err: non-nil if YAML parsing fails
+func ParseCompositeActionURLs(data []byte) (refs []ActionRef, isComposite bool, err error) {
+	var af actionFile
+	if err := yaml.Unmarshal(data, &af); err != nil {
+		return nil, false, fmt.Errorf("failed to parse action.yml: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/infrastructure/depparser/ghaworkflow/parser.go]: Fixed in 7676b23.

File: docs/adr/0008-transitive-composite-action-scanning.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+ok       action-transitive   github.com/some/transitive-action Active      No
+
+--- PURL 1 (direct) ---
+📦 Package: pkg:npm/express@4.18
+⚖️  Result: Active
+...
+
+--- PURL 6 (action-transitive) ---
+🔗 Via: github.com/some/direct-action
+📦 Package: github.com/some/transitive-action
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply docs/adr/0008-transitive-composite-action-scanning.md]: Fixed in 6ad451b — updated ADR example to show Package before Via, matching the actual `printAnalysisBody` implementation.

**Comment by kotakanbe** [comment]
@claude Follow the procedure in `.github/prompts/review.prompt.md` with `--copilot-only` mode for this PR.

Phase 2: Resolve all unresolved Copilot review threads (discover → classify → fix → build/test/lint → commit/push → reply/resolve).
Phase 3: Detect recurring patterns from FIX-classified comments, propose and apply coding rules to `.github/instructions/`, then run `make sync-instructions`.
Phase 4: Post a summary comment on this PR.
